examples/tokenize/tokenize.cpp

```diff
@@ -3,42 +3,439 @@
+//  strings, as an STL vector<string>.
+//
+//  In particular, it handles character encoding shenanigans on Windows.
+//
+// Note: raw_argc and raw_argv are not actually read at all on Windows.
+//       On Windows we call GetCommandLineW to get the arguments in wchar_t
+//       format, ignoring the regular argc/argv arguments to main().
+//
+// TODO: potential opportunity to roll common stuff into common/console.cpp
+//       in relation to Windows wchar_t shenanigans.
+static std::vector<std::string> ingest_args(int raw_argc, char ** raw_argv) {
+    std::vector<std::string> argv;
+
+    // Handle Windows, if given non-ASCII arguments.
+    // We convert wchar_t arguments into UTF-8 char* on this platform.
+    // Lets you invoke 'tokenize' on Windows cmd.exe with non-ASCII characters
+    // without throwing tantrums.
+#if defined(_WIN32)
+    int argc;
+    const LPWSTR cmdline_wargv = GetCommandLineW();
+    LPWSTR * wargv = CommandLineToArgvW(cmdline_wargv, &argc);
+
+    // silence unused arg warnings
+    (void) raw_argc;
+    (void) raw_argv;
+
+    for (int i = 0; i < argc; ++i) {
+        int length_needed = WideCharToMultiByte(CP_UTF8, 0, wargv[i], wcslen(wargv[i]), 0, 0, NULL, NULL);
+        char * output_buf = (char *) calloc(length_needed+1, sizeof(char));
+        GGML_ASSERT(output_buf);
+
+        WideCharToMultiByte(CP_UTF8, 0, wargv[i], wcslen(wargv[i]), output_buf, length_needed, NULL, NULL);
+        output_buf[length_needed] = '\0';
+
+        argv.push_back(output_buf);
+        free(output_buf);
+    }
+
+    LocalFree((HLOCAL) wargv);
+#else
+    int argc = raw_argc;
+    for (int i = 0; i < argc; ++i) {
+        argv.push_back(raw_argv[i]);
+    }
+#endif
+
+    GGML_ASSERT((unsigned int) argc == argv.size());
+
+    return argv;
+}
+
+//
+// Function: write_utf8_cstr_to_stdout(const char *) -> <writes to stdout>
+//
+// writes a string to standard output; taking into account that on Windows
+// to display correctly you have to use special handling. Works even if the
+// user has not set a unicode code page on a Windows cmd.exe.
+//
+// In case of invalid UTF-8, invalid_utf8 is set to true on Windows, and something
+// a human-readable is written instead.
+//
+// On non-Windows systems, simply printfs() the string.
+static void write_utf8_cstr_to_stdout(const char * str, bool & invalid_utf8) {
+        invalid_utf8 = false;
+
+#if defined(_WIN32)
+        // Are we in a console?
+        HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
+        DWORD dwMode = 0;
+
+        // According to Microsoft docs:
+        // "WriteConsole fails if it is used with a standard handle that is redirected to a file."
+        // Also according to the docs, you can use GetConsoleMode to check for that.
+        if (hConsole == INVALID_HANDLE_VALUE || !GetConsoleMode(hConsole, &dwMode)) {
+            printf("%s", str);
+            return;
+        }
```

Comment:
Falling back to WriteFileA just because you can't WriteConsoleW isn't ideal. Why not use the standard wprintf() so it works regardless of redirection?

Reply:
When I was testing this, it correctly piped non-ASCII text through the program on Windows. I think that condition (console reading failing) should only happen if you are piping. But not entirely sure; I don't typically develop on Windows so I could totally be wrong here.

I'm not that familiar with Windows but I got the sense that simple `printf` will just encoding-agnostically let bytes through in a pipe scenario where you are not involving the console. Do you know if that's correct? I actually looked at `WriteFile` first, but I couldn't tell if it was really any different from just doing a simpler `printf`, and thought a simple `printf` is much simpler. I didn't see if `WriteFileA` or `WriteFileW` even exist so I assumed `WriteFile` does not care about encodings, and is purely byte-oriented.

All the complicated Windows stuff is either about 1) handling command line argument encoding 2) rendering text correctly when it comes out on other side. (this function is 2)

Reply:
Actually...I just realized I did not test the output pipe. Only input pipe, i.e. stdin. Okay, so that might definitely be wrong if printf doesn't work propely.

I'll come back to test this. If it doesn't work, I'll replace it.

Reply:
Sorry, there is no WriteFileA, but I wouldn't be surprised if printf() to a stream open in text mode is going to do a conversion from the system code page (e.g. Windows-1252) to UTF-16 internally.

Reply:
Just tested it; I _think_ it's doing the same thing `WriteFile` would in this situation, i.e. it seems to behave (almost) encoding-agnostic with pipes. It does seem to turn `\n` into `\r\n`. Messing with `chcp` does not seem to affect results, got same md5sum for all outputs and it looks the same.

<img width="866" alt="command_example2" src="https://github.com/ggerganov/llama.cpp/assets/833719/0da67a86-4813-4738-912d-4241e81fc2e5">

When doing `cat` on Mac they look identical; but the Windows one has `\r\n` instead of just `\n` so it's following Windows style line endings (not visible in the screenshot below but you can see it in `hexedit` or have `vim` tell you). (out.txt = made with `tokenize` on mac, `out_windows.txt` = the one from above screenshots moved to Mac).

<img width="677" alt="Screenshot 2024-03-26 at 3 22 56 PM" src="https://github.com/ggerganov/llama.cpp/assets/833719/11be36d9-70f6-4a97-ae73-ab225eaf2e9a">

I checked with an extra `fprintf(stderr, ...)` in the code branch to check it's actually running the `printf`.

---

> Why not use the standard wprintf() so it works regardless of redirection?

When I was Googling around how to put Unicode stuff visible in `cmd.exe` no matter the code page reliably, I got the sense that `WriteConsoleW` and `wprintf` are not exactly the same, the former being less likely to do anything weird, since it works directly with the console itself. And you'd have to use `_setmode` with `wprintf`. But I admit didn't test the the wprintf approach, just went with the first implementation that worked which was the `WriteConsoleW` one.